[IMP] web: domain selector: add (not) any operator

pipu-odoo opened the pull request:
In this commit, two news operators are added in domain selector :
["ANY", "NOT ANY"].
The path of (not) any operator is a many2x or one2many field.
The value of these two operators has to be a Domain.
Note that the subdomain can contains itself a domain 
as value when the operator is (not) any.

e.g.
```
[("product_id", "any",
	["&", ("color", "in", ["red", "blue", "green"]), ("name", "ilike", "bo")]
)]
```

task~3599274

https://github.com/odoo/enterprise/pull/58871

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/143949).

pipu-odoo commented:
Explaination of modifications : 
- renamed formatValue used in leafToString to avoid abigous function name with formatValue of condition_tree.js.
- removed export of functions when it's not applicable
- simplified the utilisation of useGetDomainTreeDescription()

Polymorphe57 commented on addons/web/static/src/core/tree_editor/tree_editor_value_editors.js:
```diff
@@ -296,6 +332,34 @@ export function getValueEditorInfo(fieldDef, operator, options = {}) {
+function getDefaultCondition(fieldDefs) {
+    const defaultPath = getDefaultPath(fieldDefs);
+    const fieldDef = fieldDefs[defaultPath];
+    const operator = getDomainDisplayedOperators(fieldDef)[0];
```
this one should not be here?

Polymorphe57 commented on addons/web/static/src/core/tree_editor/utils.js:
```diff
@@ -111,16 +168,108 @@ export function leafToString(tree, fieldDef, displayNames) {
+ * @param {Boolean} isSubExpression
+ * @returns {string}
+ */
+async function getDomainTreeDescription(
```
-> getTreeDesciption?

pipu-odoo replied:
yup

Polymorphe57 commented on addons/web/static/src/views/fields/domain/domain_field.xml:
```diff
@@ -1,42 +1,46 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <templates xml:space="preserve">
 
+	<t t-name="web.DomainField.facets">
```
should it be a template with another name and put elsewhere?

Polymorphe57 commented on addons/web/static/src/views/fields/domain/domain_field.xml:
```diff
@@ -1,42 +1,46 @@
 <templates xml:space="preserve">
 
+	<t t-name="web.DomainField.facets">
+		<t t-if="!state.facets.length">
```
could be facets
and facets be defined it t-call

Polymorphe57 replied:
what about readonly? ...

Polymorphe57 replied:
props.readonly?

Polymorphe57 commented on addons/web/static/src/views/fields/domain/domain_field.xml:
```diff
@@ -1,42 +1,46 @@
-                    </div>
+						<i data-tooltip="Modify filter" class="fa fa-lg fa-caret-right pe-2"/>
+						<div class="d-print-contents">
+							<t t-call="web.DomainField.facets"></t>
```
<t t-call="web.DomainField.facets">
  <t t-set="facets" t-value="state.facets" />
</t>
?

Polymorphe57 commented on addons/web/static/src/core/tree_editor/tree_editor.xml:
```diff
@@ -147,9 +147,13 @@
         <t t-set="description" t-value="getDescription(node)" />
         <div class="o_tree_editor_condition d-flex gap-1 px-2 border bg-100">
-            <ModelFieldSelector path="node.path" resModel="props.resModel" readonly="true" />
+            <ModelFieldSelector path="node.fullPath" resModel="props.resModel" readonly="true" />
```
should be refactored at some point

pipu-odoo commented on addons/web/static/src/core/tree_editor/tree_editor_operator_editor.js:
I've checked to rollback but there are circular depedencies ...

pipu-odoo commented on addons/web/static/src/core/tree_editor/utils.js:
```diff
@@ -68,15 +77,67 @@ export function disambiguate(value, displayNames) {
+    for (const path of paths) {
+        promises.push(
+            loadPathDescription(resModel, path).then(({ displayNames }) => {
+                pathDescriptions.set(formatValue(path), displayNames.join(" \u2794 "));
```
2192 to unbold arrow

Polymorphe57 commented on addons/web/static/src/search/search_model.js:
```diff
@@ -827,8 +827,23 @@ export class SearchModel extends EventBus {
+                const node_tree = treeFromDomain(tree.value);
+                const node_field = this.searchViewFields[tree.path];
+                const node_model = getResModel(node_field);
+                description.push("[");
```
only one push?

Polymorphe57 commented on addons/web/static/tests/core/domain_selector_tests.js:
```diff
@@ -2262,4 +2285,145 @@ QUnit.module("Components", (hooks) => {
+        const parent = await makeDomainSelector({ readonly: true });
+        for (const { domain, text } of toTest) {
+            await parent.set(domain);
+            assert.strictEqual(target.querySelector(".o_domain_selector")?.innerText, text);
```
add a test with [|, (..., any, ...), ( ...)]
(not working I guess)

Polymorphe57 commented on addons/web/static/src/core/tree_editor/tree_editor_value_editors.js:
```diff
@@ -14,9 +14,16 @@ import {
+    Select,
+    List,
+    Range,
+    DomainFacets,
```
find another name?

Polymorphe57 commented on addons/web/static/src/core/tree_editor/tree_editor_operator_editor.js:
```diff
@@ -1,44 +1,10 @@
-import { sprintf } from "@web/core/utils/strings";
+import { formatValue, toValue } from "@web/core/tree_editor/condition_tree";
 import { parseExpr } from "@web/core/py_js/py";
 import { Select } from "@web/core/tree_editor/tree_editor_components";
```
put DomainFacets in another file to avoid cycle in dependencies and do not move these utils

Polymorphe57 commented on addons/web/static/src/core/tree_editor/tree_editor.xml:
```diff
@@ -3,10 +3,10 @@
     <t t-name="web.TreeEditor">
         <div class="o_tree_editor w-100" aria-atomic="true" t-att-class="className">
-            <div t-attf-class="o_tree_editor_node d-flex flex-column #{props.readonly ? 'gap-2' : 'gap-1'}">
+            <div t-attf-class="o_tree_editor_node d-flex flex-column #{props.readonly ? (props.subtree ? 'gap-0' : 'gap-2') : 'gap-1'}">
```
props.isSubTree

Polymorphe57 commented on addons/web/static/src/core/tree_editor/tree_editor.xml:
```diff
@@ -145,11 +145,14 @@
-            <ModelFieldSelector path="node.path" resModel="props.resModel" readonly="true" />
-            <span class="fst-italic" t-esc="description.operatorDescription" />
-            <t t-if="description.valueDescription">
+			<div class="fw-bolder text-nowrap" t-esc="description.pathDescription" />
```
not the same html as before (see model_field_selector.xml)
--> also, shouldn't we use the path editor from getPathEditorInfo in readonly?

Polymorphe57 commented on addons/web/static/src/core/domain_selector/domain_selector.js:
```diff
@@ -65,18 +68,18 @@ export class DomainSelector extends Component {
-        const paths = new Set([...extractPathsFromDomain(domain), "active"]);
-        await this.loadFieldDefs(p.resModel, paths);
+        const tree = treeFromDomain(domain);
+        this.fieldDefs = await loadFieldDefs(p.resModel, this.fieldService, tree, ["active"]);
```
weird API
Also why not a hook like useLoadFieldInfo

Polymorphe57 commented on addons/web/static/src/search/search_model.js:
```diff
@@ -827,8 +827,20 @@ export class SearchModel extends EventBus {
+        const getDescription = async (model, tree) => {
+            const description = [];
+            description.push(await this.getDomainTreeDescription(model, tree));
+            if (isTree(tree.value)) {
```
That part should be integretated to getDomainTreeDescription
It is missing in domain_field.js

Polymorphe57 replied:
that part does not work for a or with any inside

pipu-odoo replied:
yup. I modified it

Polymorphe57 reviewed (commented):
matchCondition (domain.js) should be adapted + test

pipu-odoo commented on addons/web/static/src/search/search_model.js:
```diff
@@ -827,8 +827,20 @@ export class SearchModel extends EventBus {
+            const description = [];
+            description.push(await this.getDomainTreeDescription(model, tree));
+            if (isTree(tree.value)) {
+                const node_field = this.searchViewFields[tree.path];
```
changed

pipu-odoo commented on addons/web/static/src/core/tree_editor/tree_editor_operator_editor.js:
```diff
@@ -38,6 +38,9 @@ const OPERATOR_DESCRIPTIONS = {
     between: _t("is between"),
+
+    any: _t("match"),
+    "not any": _t("not match"),
```
match none of is better :)

pipu-odoo commented on addons/web/static/src/core/tree_editor/tree_editor_value_editors.js:
```diff
@@ -163,6 +163,21 @@ function getPartialValueEditorInfo(fieldDef, operator, params = {}) {
+                        component: null,
+                        extractProps: null,
+                        isSupported: (value) => true,
+                        defaultValue: () => false,
```
Changed

pipu-odoo commented on addons/web/static/src/core/domain_selector/domain_selector.js:
```diff
@@ -41,12 +43,14 @@ export class DomainSelector extends Component {
         this.tree = null;
         this.defaultCondition = null;
-        this.fieldDefs = {};
+        this.fieldDefs = new Map();
```
you modified it. By the way, why is it a problem to use a map instead of object ?

pipu-odoo commented on addons/web/static/src/core/domain_selector/domain_selector.js:
```diff
@@ -106,25 +107,17 @@ export class DomainSelector extends Component {
 
-    getPathEditorInfo() {
-        const { resModel, isDebugMode } = this.props;
+    getPathEditorInfo(resModel) {
```
no idea ...

pipu-odoo commented on addons/web/static/src/core/expression_editor/expression_editor.js:
```diff
@@ -63,19 +63,18 @@ export class ExpressionEditor extends Component {
     }
 
-    getPathEditorInfo() {
+    getPathEditorInfo(resModel) {
```
if(resModel !== this.props.resModel) => throw error

Polymorphe57 commented on addons/web/static/tests/legacy/core/domain_selector_tests.js:
```diff
@@ -2272,4 +2304,114 @@ QUnit.module("Components", (hooks) => {
+        const toTest = [
+            {
+                domain: `[("product_id", "any", [("machin", "in", ["chose", "truc"] )] )]`,
+                text: `Match\nall\nof the following rules:\nProduct\nmatch\nall\nof:\nmachin\nis in\n(\nchose\n,\ntruc\n)`,
```
should be ...Product\nmatches...

Polymorphe57 commented on addons/web/static/tests/legacy/core/domain_selector_tests.js:
```diff
@@ -2272,4 +2304,114 @@ QUnit.module("Components", (hooks) => {
+            domain: `[("player_ids", "any", [('name', 'in', ["Kevin De Bruyne", "Jeremy Doku"])])]`,
+            readonly: true,
+        });
+        const text = `Match\nall\nof the following rules:\nPlayers\nmatches\nall\nof:\nPlayer Name\nis in\n(\nKevin De Bruyne\n,\nJeremy Doku\n)`;
```
should be ...Players\nmatch...

Polymorphe57 commented on addons/web/static/src/core/tree_editor/utils.js:
```diff
@@ -109,16 +198,53 @@ export function leafToString(tree, fieldDef, displayNames) {
+            }
+            return stringDescription.join(" ");
+        }
+        return await getTreeDescription(resModel, tree);
```
await not necessary

pipu-odoo replied:
why ? ... I think so

Polymorphe57 commented:
should be rebased

pipu-odoo commented on addons/web/static/src/core/domain_selector/domain_selector.js:
```diff
@@ -104,25 +105,17 @@ export class DomainSelector extends Component {
-    getPathEditorInfo() {
-        const { resModel, isDebugMode } = this.props;
+    getPathEditorInfo(resModel) {
+        const { isDebugMode } = this.props;
```
m2o : resModel1 -> resModel2
resModel1 has a field country_id that will be used as defaultPath
resModel2 hasn't country_id
make code call getPathEditorInfo -> crash
In the test have a any operator then add a rule in subDomain

Polymorphe57 commented on addons/web/static/src/core/domain_selector/domain_selector.js:
```diff
@@ -63,27 +66,25 @@ export class DomainSelector extends Component {
+        const tree = treeFromDomain(domain);
 
-        const [defaultCondition] = await Promise.all([
+        const [defaultCondition, getFieldDef] = await Promise.all([
```
simplify

Polymorphe57 commented on addons/web/static/src/core/expression_editor/expression_editor.js:
```diff
@@ -1,25 +1,11 @@
-function getDefaultCondition(fieldDefs) {
-    const defaultPath = getDefaultPath(fieldDefs);
-    const fieldDef = fieldDefs[defaultPath];
-    const operator = getExpressionDisplayedOperators(fieldDef)[0];
```
we missed something:
getExpressionDisplayedOperators is used instead of getDomainDisplayedOperators
and getDefaultCondition is called on this.filteredFields.
The first problem can be handled by passing the prop getDefaultOperator.
The second on can maybe solved by checking the type of fieldDef in getDefaultPath and only accept m2o fields?

Polymorphe57 replied:
this means you'll have to rework core/domain_selector/utils.js in the end I think.

Polymorphe57 replied:
or maybe we can juste ignore the param defaultCondition in getPathEditorInfo and compute it like it was done in before in ExpressionEditor (since refuse other res models anyway)

Polymorphe57 commented:
@robodoo r+

robodoo commented:
@pipu-odoo @polymorphe57 linked pull request(s) odoo/enterprise#58871 not ready. Linked PRs are not staged until all of them are ready.
